File: 03_forward_rates_and_fra_pricing/python/curve.py
import numpy as np


class YieldCurve:
    """
    Simple continuously compounded zero curve.

    This class provides:
    - zero rates
    - discount factors
    - implied forward rates

    FO Perspective
    --------------
    In real trading systems, curves are built from:
    - OIS swaps
    - IRS swaps
    - futures
    - FRA instruments

    Here we use simplified market zero rates.
    """

    def __init__(self, maturities, zero_rates):

        self.maturities = np.array(maturities)
        self.zero_rates = np.array(zero_rates)

    def zero_rate(self, t):
        """
        Linear interpolation of zero rates.
        """

        return np.interp(
            t,
            self.maturities,
            self.zero_rates
        )

    def discount_factor(self, t):
        """
        Compute discount factor:

        DF(t) = exp(-r*t)
        """

        r = self.zero_rate(t)

        return np.exp(-r * t)

    def forward_rate(self, t1, t2):
        """
        Compute implied forward rate.

        Formula:

        F(t1,t2)
        =
        (DF(t1)/DF(t2)-1)/(t2-t1)
        """

        df1 = self.discount_factor(t1)
        df2 = self.discount_factor(t2)

        return (
            (df1 / df2) - 1
        ) / (t2 - t1)

File: 10_short_rate_models_with_stochastic_volatility/cpp/curve.cpp
#include "curve.hpp"
#include <cmath>

YieldCurve::YieldCurve(
    const std::vector<double>& mats,
    const std::vector<double>& rates
){

    maturities = mats;
    zero_rates = rates;
}

double YieldCurve::zero_rate(double t){

    for(size_t i=1;i<maturities.size();++i){

        if(t <= maturities[i]){

            double t1 = maturities[i-1];
            double t2 = maturities[i];

            double r1 = zero_rates[i-1];
            double r2 = zero_rates[i];

            double w =
                (t-t1)/(t2-t1);

            return r1 + w*(r2-r1);
        }
    }

    return zero_rates.back();
}

double YieldCurve::discount_factor(double t){

    double r = zero_rate(t);

    return exp(-r*t);
}

double YieldCurve::forward_rate(
    double t1,
    double t2
){

    double df1 =
        discount_factor(t1);

    double df2 =
        discount_factor(t2);

    return
        ((df1/df2)-1.0)/(t2-t1);
}
